Implement unsubscribe flow

matthewheath opened the pull request:
This PR implements #190 and fixes some over-eager find and replace mistakes in the test suite. It uses the new settings store to handle email notification preferences.

It also includes the commits for `convert-settings-to-store` due to a branch snafu but these can be ignored. Additionally, the existing `new_user` entry in the `Users.yml` fixture conflicted with the `new_user` method in the User model spec so it has been renamed.

sudara commented:
@matthewheath Wow, this is really cool!!!! 

Do we also have to add the unsubscribe link to the notification email going out?

Also, #190 also addresses a "unfollow user" link (following a user results in upload notifications being emailed out if you allow emails) so we should either create a new issue to handle "instant unfollow" links or keep that issue open until that is attacked as well. Sorry if that was unclear!

Really thankful for the help on this!

matthewheath commented:
@sudara 

1. Yes, I forgot to add the unsubscribe link to the email. I'll push another commit to this PR.

2. Right now disabling emails like this also appears to disable upload notifications as those behaviours seem to rely on `wants_email?` in the `User` model. Which would be a better flow—unfollowing everyone as part of this `unsubscribe` action or performing that action separately?

sudara commented:
@matthewheath Hey! So, sorry for the lack of clarity. Here's how things work currently:

1. I presses "follow" on Alice (right now it's just the heart button on the /user page, that's being fixed shortly)
2. Now that I'm following Alice, when she uploads, I'll get an email
3. I also follow Bob, so when he uploads, I'll get email.
4. If I tire of getting *any* email, then I want to hit "unsubscribe from all upload notification emails" (implemented in this PR)
5. If I like getting the emails, but I'm tired of following bob, then it would be nice to separately have an option to "Unfollow Bob to stop getting notified on Bob's uploads" (not yet implemented, but what #190 was trying to get at)

matthewheath commented:
@sudara I've now implemented the unfollow feature as intended — can you review the [two issues](https://codeclimate.com/github/sudara/alonetone/pull/246) raised by CodeClimate to help me refactor them?

I've also implemented the unfollow and unsubscribe links in the relevant mailers, and added test suites to cover all commits.

sudara commented on app/controllers/following_controller.rb:
```diff
@@ -0,0 +1,45 @@
+class FollowingController < ApplicationController
+  before_action :require_login, :redirect_if_current_user_equals_params
```
If you want to please code climate, in addition to `require_login` you could use `find_user` and then have @user set for you in the action.

sudara commented on app/controllers/following_controller.rb:
```diff
@@ -0,0 +1,45 @@
+  def follow
+    followee = User.find_by!(login: params[:login])
+
+    if already_following?(followee.id)
```
With `@user` set, then you could drop the `find_by! and even the `already_following` method and go with `if current_user.is_following? @user`

matthewheath commented:
@sudara I've implemented all of the suggested refactors.

sudara commented:
@matthewheath Awesome! Merged, but seeing some CI fails on master, maybe we need to stub out the API call...

https://travis-ci.com/sudara/alonetone/builds/87705752

```
Failures:
  1) CommentsController basics should allow private comment on track
     Failure/Error: self.is_spam = spam? # makes API request
     Net::OpenTimeout:
       execution expired
     # ./app/models/comment.rb:51:in `set_spam_status'
     # ./app/controllers/comments_controller.rb:9:in `create'
     # ./spec/controllers/comments_controller_spec.rb:35:in `block (3 levels) in <top (required)>'
  2) CommentsController basics should allow user comment on a track (via xhr)
     Failure/Error: self.is_spam = spam? # makes API request
     Net::OpenTimeout:
       execution expired
     # ./app/models/comment.rb:51:in `set_spam_status'
     # ./app/controllers/comments_controller.rb:9:in `create'
     # ./spec/controllers/comments_controller_spec.rb:28:in `block (3 levels) in <top (required)>'
```

matthewheath commented:
That's weird @sudara when I ran the tests locally they all passed. Maybe the API was down then? I think we should probably be stubbing that external call anyway, yes.
